pkg/internal/ingress/convert.go

```diff
@@ -194,7 +207,20 @@ func autoConvert_networking_IngressBackend_To_v1beta1_IngressBackend(in *network
 }
 
 func Convert_v1beta1_Ingress_To_networking_Ingress(in *networkingv1beta1.Ingress, out *networkingv1.Ingress, s conversion.Scope) error {
-	return autoConvert_v1beta1_Ingress_To_networking_Ingress(in, out, s)
+	err := autoConvert_v1beta1_Ingress_To_networking_Ingress(in, out, s)
+	if err != nil {
+		return err
+	}
+	// v1beta1 Ingresses should not have IngressClassName set but instead use the deprecated annotation.
+	// Move the ingress class from the annotations to the Spec
+	if in.Annotations == nil {
+		return nil
+	}
+	if ingressClass, found := in.Annotations["kubernetes.io/ingress.class"]; found {
+		out.Spec.IngressClassName = &ingressClass
+		delete(in.Annotations, "kubernetes.io/ingress.class")
```

Comment:
question: is it usual for these conversion functions to mutate `in` like this by removing the annotation? on the face of it that seems surprising to me; is it to guard against modifications to `in` down the line changing `out.Spec.IngressClassName` since we take `&ingressClass`?

If so, would it work to copy the string and avoid mutating `in` here? I don't think that'd be required, though

Reply:
The conversion functions don't really convert, but use unsafe.Pointer manipulation to make the new object and old object point to the same underlying data. So in this case manipulating out.Annotations and in.Annotations manipulates the same thing. 

When I use these functions I [deepcopy `in` first](https://github.com/jetstack/cert-manager/blob/aa19969bf241cf9cc2688d00a202a2cb83fbd210/pkg/internal/ingress/v1beta1.go#L97), so mutating the Object is safe (although this is not at all obvious and a probably footgun for future maintainers)

Would you suggest I always deepcopy inside the convert function? Or should I make the documentation better?

Reply:
Thinking about it more, I guess what I'm trying to work out is why we delete the annotation on `in` at all; if in practice (as in [your link](https://github.com/jetstack/cert-manager/blob/aa19969bf241cf9cc2688d00a202a2cb83fbd210/pkg/internal/ingress/v1beta1.go#L97)) we're using a deep-copied object (to which we keep no reference I can see), why do we care what in `in`'s annotations after we're done with the conversion?

I can see why we'd make sure that such an annotation doesn't end up in `out`, but not `in`.

As for your suggestions: I'm unlikely to ever say no to additional comments explaining what's going on :joy:  But I think there's enough of a "here be dragons" vibe to this whole package / conversion workflow that it isn't really needed.

If we always deepcopy `in` at call sites, it makes sense to me to do the deepcopy in the conversion function just to remove a footgun - but again, I don't think that's required at all and it's probably not worth the time to make the change.

Reply:
in.Annotations and out.Annotations are the same map, but it was not obvious so I made it more explicit.
I added more comments explaining exactly what is happening, and additional tests to cover all code paths.